_posts/2020-09-14-notes-on-building-python-for-windows-using-rbm.md

```diff
@@ -0,0 +1,77 @@
+---
+layout: post
+title: "Notes on building Python for Windows using RBM"
+author: yanmaani
+tags: [News]
+---
+
+Namecoin would like for the Tor Browser to support .bit domains.
+The only mature, lightweight way to do this is to use the wallet qua name resolver Electrum-NMC.
+It is written in Python.
+To run Python software, you need an interpreter, like CPython.
+
+Namecoin and Tor are both intended to be secure projects.
+All their binaries have to build reproducibly.
+The Tor Browser is also intended to be portable.
+It's not acceptable for users to have to install additional software on their computer to run the Tor Browser.
+
+This means a Python interpreter would have to be bundled in order for Electrum-NMC to be includible in the Tor Browser.
+And since the Tor Browser is a secure project, said Python interpreter has to be built reproducibly.
+For this reason, the `tor-browser-build` repository has received some patches to reproducibly build such binaries.
+
+This post details the problems encountered while making them.
+
+Building CPython reproducibly for Linux is trivial.
+Download the source code, patch it to disable build timestamps, and compile.
+That's it, you're done.
+
+The situation is markedly different on Windows.
+The official Python documentation suggests [building with Microsoft Visual Studio 2017](https://github.com/python/cpython/blob/master/PCbuild/readme.txt).
+This is no good.
+That compiler doesn't support reproducible builds, it's closed-source, and it would have to run in Wine.
+
+There should be another way.
+CPython is written in standard C.
+GCC can cross-compile C software for Windows using MinGW - the compiler runs on Linux, but produces binaries for Windows.
+That's how the Tor Project build Tor Browser for Windows.
+In theory, it should be possible to use GCC to cross-compile CPython for Windows, and a lot of Tor's tooling should be possible to re-use.
+
+In practice, Python's ordinary build system doesn't support this.
+It isn't made for cross-compilation.
+Thankfully, one [Erik Janssens](https://github.com/erikjanss) (no relation to the Namecoin project) had created an alternative [Meson](https://mesonbuild.com/)-based [build script](https://github.com/v-finance/cross-python) to compile Python for Windows with Linux using GCC and MinGW.
```

Comment:
I don't see much reason to clarify that Erik is unaffiliated with us; we regularly cite external work in news posts, and we've never had a need to clarify this.

Reply:
Amended in part.